mcproto/types/nbt.py

```diff
@@ -0,0 +1,1315 @@
+
+        :return: The NBT tag.
+        """
+        return cls.deserialize(buf, with_name=with_name, with_type=with_type)
+
+    @staticmethod
+    def from_object(data: object, /, name: str = "", *, use_int_array: bool = True) -> NBTag:  # noqa: PLR0911,PLR0912
+        """Create an NBT tag from an arbitrary (compatible) Python object.
+
+        :param data: The object to convert to an NBT tag.
+        :param use_int_array: Whether to use IntArrayNBT and LongArrayNBT for lists of integers.
+            If set to False, all lists of integers will be considered as ListNBT.
+        :param name: The name of the resulting tag. Used for recursive calls.
+
+        :return: The NBT tag representing the object.
+
+        :note: The function will attempt to convert the object to an NBT tag in the following way:
+            - If the object is a dictionary with a single key, the key will be used as the name of the tag.
+            - If the object is an integer, it will be converted to a ByteNBT, ShortNBT, IntNBT, or LongNBT tag
+                depending on the value.
+            - If the object is a list, it will be converted to a ListNBT tag.
+            - If the object is a dictionary, it will be converted to a CompoundNBT tag.
+            - If the object is a string, it will be converted to a StringNBT tag.
+            - If the object is a float, it will be converted to a FloatNBT tag.
+            - If the object can be serialized to bytes, it will be converted to a ByteArrayNBT tag.
+
+
+            - If you want an object to be serialized in a specific way, you can implement:
+
+            ```python
+            def to_nbt(self, name: str = "") -> NBTag:
+                ...
+            ```
+        """
+        if hasattr(data, "to_nbt"):  # For objects that can be converted to NBT
+            return data.to_nbt(name=name)  # type: ignore
+
+        if isinstance(data, int):
+            if -(1 << 7) <= data < 1 << 7:
+                return ByteNBT(data, name=name)
+            if -(1 << 15) <= data < 1 << 15:
+                return ShortNBT(data, name=name)
+            if -(1 << 31) <= data < 1 << 31:
+                return IntNBT(data, name=name)
+            if -(1 << 63) <= data < 1 << 63:
+                return LongNBT(data, name=name)
+            raise ValueError(f"Integer {data} is out of range.")
+        if isinstance(data, float):
+            return FloatNBT(data, name=name)
+        if isinstance(data, str):
+            return StringNBT(data, name=name)
+        if isinstance(data, (bytearray, bytes)):
+            if isinstance(data, bytearray):
+                data = bytes(data)
+            return ByteArrayNBT(data, name=name)
+        if isinstance(data, list):
+            if not data:
+                # Type END is used to mark an empty list
+                return ListNBT([], name=name)
+            first_type = type(data[0])
+            if any(type(item) != first_type for item in data):
```

Comment:
comparing 2 types (classes) should probably be done with `is not` rather than `!=`.